src/pycsou/opt/solver/pds.py

```diff
@@ -0,0 +1,965 @@
+                ]
+            )
+            raise ValueError(msg)
+
+    @pycrt.enforce_precision(i=["x0", "z0", "tau", "sigma", "rho", "beta"], allow_None=True)
+    def m_init(
+        self,
+        x0: pyct.NDArray,
+        z0: typ.Optional[pyct.NDArray],
+        tau: typ.Optional[pyct.Real] = None,
+        sigma: typ.Optional[pyct.Real] = None,
+        rho: typ.Optional[pyct.Real] = None,
+        beta: typ.Optional[pyct.Real] = None,
+    ):
+        mst = self._mstate  # shorthand
+        mst["x"] = mst["x_prev"] = x0
+        mst["z"] = mst["z_prev"] = self.set_dual_variable(z0)
+        mst["beta"] = self.set_beta(beta)
+        mst["tau"], mst["sigma"] = self.set_step_sizes(tau, sigma)
+        mst["rho"] = self.set_momentum_term(rho)
+
+    def m_step(self):
+        mst = self._mstate  # shorthand
+        x_temp = self._g.prox(
+            mst["x"] - mst["tau"] * self._f.grad(mst["x"]) - mst["tau"] * self._K.adjoint(mst["z"]),
+            tau=mst["tau"],
+        )
+        if not isinstance(self._h, pyclo.NullFunc):
+            u = 2 * x_temp - mst["x"]
+            z_temp = self._h.fenchel_prox(mst["z"] + mst["sigma"] * self._K(u), sigma=mst["sigma"])
+            mst["z"] = mst["rho"] * z_temp + (1 - mst["rho"]) * mst["z"]
+        mst["x"] = mst["rho"] * x_temp + (1 - mst["rho"]) * mst["x"]
+
+    def default_stop_crit(self) -> pycs.StoppingCriterion:
+        stop_crit_x = pycos.RelError(
+            eps=1e-4,
+            var="x",
+            f=None,
+            norm=2,
+            satisfy_all=True,
+        )
+        stop_crit_z = pycos.RelError(
+            eps=1e-4,
+            var="z",
+            f=None,
+            norm=2,
+            satisfy_all=True,
+        )
+        return stop_crit_x & stop_crit_z
+
+    def solution(self) -> pyct.NDArray:
+        """
+        Returns
+        -------
+        p: NDArray
+            (..., N) x solution.
+        d: NDArray
+            (..., N) z solution.
+        """
+        data, _ = self.stats()
+        return data.get("x")
+
+    def set_beta(self, beta: typ.Optional[pyct.Real]) -> float:
+        r"""
+        Sets the Lipschitz constant.
+        Returns
+        -------
+        float
+            Lipschitz constant.
+        """
+        if beta is None:
+            if math.isfinite(dl := self._f._diff_lipschitz):
+                return pycrt.coerce(dl)
+            else:
+                msg = "beta: automatic inference not supported for operators with unbounded Lipschitz gradients."
+            raise ValueError(msg)
+        else:
+            try:
+                assert beta > 0
+                return pycrt.coerce(beta)
+            except:
+                raise ValueError(f"beta must be positive, got {beta}.")
+
+    def set_step_sizes(self, tau: typ.Optional[pyct.Real], sigma: typ.Optional[pyct.Real]) -> typ.Tuple[float, float]:
+        r"""
+        Set the x/z step sizes.
+        Returns
```

Comment:
you should always have blank line after the 1st line of docstrings (this way Sphinx interprets the first line as the short description of the function).

Reply:
More generally always put blank lines after/before functions or the formatting will be wrong.